src/JustSaying/Fluent/PublishConfig/DynamicMessagePublisher.cs

```diff
@@ -83,37 +53,27 @@ public async Task PublishAsync(IEnumerable<Message> messages, PublishBatchMetada
             {
                 string topicName = groupByTopic.Key;
                 var batch = groupByTopic.ToList();
-                IMessageBatchPublisher publisher;
-                if (_batchPublisherCache.TryGetValue(topicName, out publisher))
-                {
-                    publisherTask.Add(publisher.PublishAsync(batch, metadata, cancellationToken));
-                    continue;
-                }
-
-                var lockObj = _topicCreationLocks.GetOrAdd(topicName, _ => new SemaphoreSlim(1, 1));
-                _logger.LogDebug("Publisher for topic {TopicName} not found, waiting on creation lock", topicName);
-                using (await lockObj.WaitScopedAsync(cancellationToken).ConfigureAwait(false))
-                {
-                    if (_batchPublisherCache.TryGetValue(topicName, out publisher))
-                    {
-                        _logger.LogDebug("Lock re-entrancy detected, returning existing publisher");
-                        publisherTask.Add(publisher.PublishAsync(batch, metadata, cancellationToken));
-                        continue;
-                    }
-
-                    _logger.LogDebug("Lock acquired to initialize topic {TopicName}", topicName);
-                    var config = _staticConfigBuilder(topicName);
-                    _logger.LogDebug("Executing startup task for topic {TopicName}", topicName);
-                    await config.StartupTask(cancellationToken).ConfigureAwait(false);
-
-                    publisher = _batchPublisherCache.GetOrAdd(topicName, config.BatchPublisher);
-                }
-
-                _logger.LogDebug("Publishing message on newly created topic {TopicName}", topicName);
-                publisherTask.Add(publisher.PublishAsync(batch, metadata, cancellationToken));
+                var publisherConfigTask = _publisherConfigurationCache.GetOrAdd(topicName, _ => CreateLazyPublisherConfig(topicName, cancellationToken)).Value;
+                var publisherConfig = await publisherConfigTask;
+                publisherTask.Add(publisherConfig.BatchPublisher.PublishAsync(batch, metadata, cancellationToken));
             }
         }
 
         await Task.WhenAll(publisherTask).ConfigureAwait(false);
     }
+
+    private Dictionary<string, InterrogationResult> GetInterrogationResultForTasks(Func<StaticPublicationConfiguration, InterrogationResult> interrogate) =>
+        _publisherConfigurationCache
+            .Where(w => w.Value.Value.IsCompleted)
+            .OrderBy(o => o.Key)
+            .ToDictionary(x => x.Key, x => interrogate(x.Value.Value.Result));
```

Comment:
This is a bit gnarly 😅

Reply:
I like it tbh 😅

Reply:
I think I should change the where to `w.Value.Value.Status == TaskStatus .RanToCompletion`, otherwise .Result could throw if it's faulted or cancelled.

Reply:
Maybe, alternatively we could just do away with the `Lazy<T>` part, `Task<T>` already has the right behaviour.

Reply:
Ah, you're right, ignore me